### DISCUSSION TITLE
ControlNet Channels - Restore missing RGB channel

### DISCUSSION THREAD

**Discussion by geroldmeisinger** [lllyasviel/ControlNet#567] [Show and tell]
I created a [ControlNet - Channels RGB](https://huggingface.co/GeroldMeisinger/controlnet-channels) (`214244a32 drop=0.5 mp=fp16 lr=1e-5 epochs=2`) which was trained on RGB images where each of them have one channel missing in order to restore a missing channel from an image (a special case of the Recolor ControlNet). You can either use the inference as a whole (but it changes the semantics a bit), or only use the missing channel from the inference (which doesn't look to good unfortunately). I also created an accompanying [space which lets you replace the channel from an inference image](https://huggingface.co/spaces/GeroldMeisinger/channels). A similar [evaluation using classic algorithms](https://codegolf.stackexchange.com/questions/79219/reconstruct-a-missing-rgb-channel) can be found on stackexchange. Of course it should be possible to extend this concept to color spaces like HSV or LAB and infer things like saturation or luminance etc.
* sudden change at: 227200a32(!) .. that's is pretty late
* sufficiently phenomenal convergence at: 320000a32

TODO: I think I need to try img2img

### Evaluation
first is ground truth, second is inference only, third is recomposing the missing channel from inference, the image on the right is ground truth with a missing channel

**Lenna (missing red)**

![lenna](https://github.com/lllyasviel/ControlNet/assets/112266044/db0da54a-2db6-4402-b454-f13d2a93880c)

inference only
![lenna_goodprompt_r_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/668aa597-cf49-4f05-bdd9-f4023dee21a9)

recomposed
![lenna_goodprompt_r_restored_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/4c6b9eff-278c-491b-9950-2ebc70918242)

**House2 (missing green)**

![house2](https://github.com/lllyasviel/ControlNet/assets/112266044/03720ebb-1204-4da7-a19b-5c1ce0ce5d50)

inference only
![house2_goodprompt_g_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/8cb167dd-077c-47a3-8338-955858fb98ee)

recomposed
![house2_goodprompt_g_restored_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/83fc45d5-0a8f-4e1c-b761-aac01176e19d)

**Dog2 (missing blue)**

![dog2](https://github.com/lllyasviel/ControlNet/assets/112266044/db0bd38f-c7c1-473a-b420-1e9597a00f95)

inference only
![dog2_goodprompt_b_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/a6b90e68-16a3-460a-8b77-cc88f6a69e08)

recomposed
![dog2_goodprompt_b_restored_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/db14cce3-e7dc-467e-812c-63d9ce7aae48)

**Forest (missing red)**

![forest](https://github.com/lllyasviel/ControlNet/assets/112266044/0c7b7ffe-66a3-49cc-92cd-aa444cc576f0)

inference only
![forest_goodprompt2_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/2df138e3-c951-4fd1-981b-6e254b70f8ed)

recomposed
![forest_goodprompt2_restored_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/f26e7859-3ad8-4ed5-a9dd-7a3c60bcc8b8)

**Glacier (missing green)**

![glacier](https://github.com/lllyasviel/ControlNet/assets/112266044/bb18c8fe-169f-41c2-9038-ac1ed1492d8d)

inference only
![glacier_goodprompt1_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/94927ba1-daba-4fe9-af73-613d2dd3e52e)

recomposed
![glacier_goodprompt1_restored_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/c0f00646-1cee-4ff2-ba8b-a268be5c6ee7)

**Fox (missing blue)**

![fox](https://github.com/lllyasviel/ControlNet/assets/112266044/ce754443-b2a0-4c6b-972b-677be145aacb)

inference only
![fox_goodprompt0_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/3544fbd0-d42e-4425-b253-e327fe8544e1)

recomposed
![fox_goodprompt0_restored_13392](https://github.com/lllyasviel/ControlNet/assets/112266044/d4fc5265-e3a2-4bc6-b91b-790692c573ce)

**Comment by nevenpavlovic** [comment]
Is it possible for you to host this space again?

**Reply by geroldmeisinger** [reply]
the huggingface space? i get some runtime error and have to look into it in the next few days. keep in mind that the space only replace the channel, it's not the full stable diffusion + controlnet stack. this you have to do locally.

**Reply by nevenpavlovic** [reply]
I have an image with two channels present. So either whole reconstruction or just getting a missing channel would work for me as I can always add it as a layer to complete the image.

**Reply by geroldmeisinger** [reply]
okay, but the space only merges the layers, it does not reconstruct it in the first place. for this you need to setup stable diffusion and download the controlnet locally anyway. I'm not able to run a whole SD installation online with a free HF account.
